cmd/restic/cmd_prune.go

```diff
@@ -56,207 +101,348 @@ func runPrune(gopts GlobalOptions) error {
+		switch {
+		case p.usedBlobs == 0 && p.duplicateBlobs == 0:
+			stats.packs.unused++
+		case packSize < repository.MinPackSize:
+			stats.packs.small++
+		case p.unusedBlobs == 0:
+			stats.packs.used++
+		default:
+			stats.packs.partlyUsed++
 		}
 
-		if hasActiveBlob {
-			continue
+		// decide what to do
+		switch {
+		case p.usedBlobs == 0 && p.duplicateBlobs == 0:
+			// All blobs in pack are no longer used => remove pack!
+			removePacks.Insert(id)
+			stats.blobs.remove += p.unusedBlobs
+			stats.size.remove += p.unusedSize
+
+		case opts.RepackTreesOnly && p.tpe != restic.TreeBlob,
+			// if this is a data pack and --repack-trees-only is set => keep pack!
+			uint(len(repackPacks)) >= opts.MaxRepackCount:
+			// also keep if number of packs to repack is reached
+			stats.packs.keep++
+
+		case opts.RepackMixed && p.tpe == restic.InvalidBlob,
+			opts.RepackSmall && packSize < repository.MinPackSize,
+			opts.RepackDuplicates && p.duplicateBlobs > 0:
+			// repack if the according flag is set!
+			repack(id, p)
+
+		case p.unusedBlobs == 0:
+			// All blobs in pack are used => keep pack!
+			stats.packs.keep++
+
+		case opts.MaxUnusedPercent == 0.0:
+			// repack all other packs if max-unusued-percent=0
+			repack(id, p)
+
+		default:
+			// all other packs are candidates for repacking
+			repackCandidates = append(repackCandidates, packInfoWithID{ID: id, packInfo: p})
 		}
 
-		removePacks.Insert(packID)
+		delete(indexPack, id)
+		bar.Report(restic.Stat{Blobs: 1})
+		return nil
+	})
+	bar.Done()
+	if err != nil {
+		return err
+	}
 
-		if !rewritePacks.Has(packID) {
-			return errors.Fatalf("pack %v is unneeded, but not contained in rewritePacks", packID.Str())
-		}
+	if len(indexPack) != 0 {
+		Warnf("There are packs referenced in the index that are not present in the repository: %v\n", indexPack)
+		return errorPacksMissing
+	}
 
-		rewritePacks.Delete(packID)
+	// if all duplicates are repacked, print out correct statistics
+	if opts.RepackDuplicates || (opts.MaxUnusedPercent == 0.0) {
+		stats.blobs.repackrm += stats.blobs.duplicate
+		stats.size.repackrm += stats.size.duplicate
 	}
 
-	Verbosef("will delete %d packs and rewrite %d packs, this frees %s\n",
-		len(removePacks), len(rewritePacks), formatBytes(uint64(removeBytes)))
+	// If unused space is restricted, check and find packs for repacking
+	if opts.MaxUnusedPercent < 100.0 {
+		maxUnusedSizeAfter := uint64(opts.MaxUnusedPercent / (100.0 - opts.MaxUnusedPercent) * float32(stats.size.used))
 
-	var obsoletePacks restic.IDSet
-	if len(rewritePacks) != 0 {
-		bar := newProgressMax(!gopts.Quiet, uint64(len(rewritePacks)), "packs rewritten")
-		obsoletePacks, err = repository.Repack(ctx, repo, rewritePacks, usedBlobs, bar)
-		if err != nil {
-			return err
+		//sort repackCandidates such that packs with highest ratio unused/used space are picked first
+		sort.Slice(repackCandidates, func(i, j int) bool {
+			return repackCandidates[i].unusedSize*repackCandidates[j].usedSize >
```

Comment:
In fact it's not the percentage of unused space (which should be the ratio "unused space / total space"), but it really is the ration "unused space / used space".

However, I don't use this ratio in the `sort` directly because it is not saved. Moreover, by using
``` unused[i] * used[j] > unused[j] * used[i] ```
instead of 
``` unused[i] / used[i] > unused[j] / used[j] ```
there is no need to deal with floats and no headache about whether to take care of division by zero (which actually should not occur as these pack files are already chosen for deletion...)

Is your point about clarifying the comment or about changing the sorting criteria?

Reply:
At first, my intuition was that we should sort the packs by percentage of unused space in a pack file. However, that sorting criteria yields exactly the same order as sorting by unused/used:
`unused[i] / total[i] > unused[j] / total[j]` `<=> unused[i] / (used[i] + unused[i]) > unused[j] / (used[j] + unused[j])` `<=> unused[i] * (used[j] + unused[j]) > unused[j] * (used[i] + unused[i])` `<=>  unused[i]*used[j] + unused[i]*unused[j] > unused[j]*used[i] + unused[j]*unused[i]` `<=> unused[i]*used[j] > unused[j]*used[i]`. As `used[i]` and `unused[i]` are limited to the values which fit into a uint32 value, the multiplication result fits into the `uint64` datatype used here. Division-by-zero is actually not a problem here as `used[i] >= pack.HeaderSize`.

So my point is about clarifying the comment a bit: it should mention that the criteria is equivalent to sorting by unused space percentage (I don't think it's necessary to explain why that is the case) and mention that `uint32*uint32<=uint64`

Reply:
You are right; another argument would be that for `unused==0` we are fine and if `unused > 0`, we can compare the inverse values `total / unused = 1 + used/unused` and subtract `1` on both sides of the inequality.

I changed the comment.